File: bitmask_and_bit_manipulation/single_number/single_number.py
from typing import *


def singleNumber(nums: List[int]) -> int:
    num = 0
    for item in nums:
        # XOR
        num ^= item
    return num


if __name__ == '__main__':
    test_list = [1, 2, 2, 3, 1]
    print(3 == singleNumber(test_list))

File: bitmask_and_bit_manipulation/single_number/single_number.java
class Solution {
    public int singleNumber(int[] nums) {
        int num = 0;
        for (int n : nums) {
            num ^= n;
        }
        return num;
    }
}

class Test {
    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.singleNumber(new int[]{2, 2, 1}));
        System.out.println(s.singleNumber(new int[]{4, 1, 2, 1, 2}));
        System.out.println(s.singleNumber(new int[]{1}));
    }
}